app/parse.py

```diff
@@ -15,8 +33,126 @@ class Product:
+        logging.FileHandler("parser.log"),
+        logging.StreamHandler(sys.stdout)
+    ],
+)
+
+
+class SingletonMeta(type):
+    _instances = {}
+
+    def __call__(cls, *args, **kwargs) -> Any:
+        if cls not in cls._instances:
+            instance = super().__call__(*args, **kwargs)
+            cls._instances[cls] = instance
+        return cls._instances[cls]
+
+
+class SeleniumBrowser(metaclass=SingletonMeta):
+    def __init__(self) -> None:
+        logging.info("Initializing driver")
+        chrome_options = Options()
+        chrome_options.add_argument("--headless")
+        chrome_options.add_argument("--disable-gpu")
+        chrome_options.add_argument("--no-sandbox")
+        chrome_options.add_argument("--disable-dev-shm-usage")
+        self.driver = webdriver.Chrome(options=chrome_options)
+
+    def get_driver(self) -> webdriver:
+        logging.info("Getting driver")
+        return self.driver
+
+    def close_driver(self) -> None:
+        logging.info("Closing driver")
+        self.driver.close()
+        del SeleniumBrowser._instances[type(self)]
+
+
+def parse_single_product(product: Tag) -> Product:
+    return Product(
+        title=product.select_one("a.title")["title"],
+        description=product.select_one(".description")
+        .text.replace("\xa0", " "),
+        price=float(product.select_one(".price")
+                    .text.replace("$", "")),
+        rating=int(len(product.select("span.ws-icon-star"))),
+        num_of_reviews=int(product.select_one(".review-count")
+                           .text.split()[0]),
+    )
+
+
+def get_products(driver: webdriver) -> List[Product]:
+    page_soup = BeautifulSoup(driver.page_source, "html.parser")
+    products = page_soup.select(".product-wrapper")
+    return [parse_single_product(product) for product in products]
+
+
+def get_products_from_page(url: str) -> List[Product]:
+    driver = SeleniumBrowser().get_driver()
+    driver.get(url)
+    logging.info(f"Getting products from page {url}")
+    time.sleep(5)
+    cookies_btn = driver.find_elements(By.CSS_SELECTOR, "button.acceptCookies")
+    if cookies_btn:
+        logging.info("Accepting cookies")
+        cookies_btn[0].click()
+
+    more_btn = driver.find_elements(
+        By.CSS_SELECTOR, "a.ecomerce-items-scroll-more"
+    )
+    if more_btn:
+        more_btn = more_btn[0]
+        more_btn_style = more_btn.get_attribute("style")
+        pbar = tqdm()
+        while "display: none" not in more_btn_style:
+            driver.execute_script("arguments[0].click();", more_btn)
+            more_btn = WebDriverWait(driver, 5).until(
+                e_cond.presence_of_element_located(
+                    (By.CSS_SELECTOR, "a.ecomerce-items-scroll-more")
+                )
+            )
+            more_btn_style = more_btn.get_attribute("style")
+            pbar.update(1)
+    return get_products(driver)
+
+
+def save_products_to_csv(products: list[Product], filename: str) -> None:
```

Comment:
The type hint `list[Product]` is only available in Python 3.9 and later. If you need to support earlier versions, consider using `List[Product]` from the `typing` module.